models/model_upload/multimodal_models/vllm-miniCPM-2.6/1/model.py

```diff
@@ -0,0 +1,304 @@
+
+    self.python_executable = sys.executable
+
+    # Start the VLLM server in a separate thread
+    vllm_server_thread = threading.Thread(target=self.vllm_openai_server)
+    vllm_server_thread.start()
+
+    # Wait for the server to be ready
+    self.server_started_event.wait()
+
+    models = self.client.models.list()
+    self.model = models.data[0].id
+
+  def vllm_openai_server(self):
+    os.path.join(os.path.dirname(__file__))
+
+    # if checkpoints section is in config.yaml file then checkpoints will be downloaded at this path during model upload time.
+    checkpoints = os.path.join(os.path.dirname(__file__), "checkpoints")
+    process = subprocess.Popen(
+        [
+            self.python_executable,
+            '-m',
+            'vllm.entrypoints.openai.api_server',
+            '--model',
+            checkpoints,
+            '--dtype',
+            'float16',
+            '--tensor-parallel-size',
+            '1',
+            '--quantization',
+            'awq',
+            '--gpu-memory-utilization',
+            '0.8',
+            '--port',
+            str(self.port),
+            '--host',
+            'localhost',
+            '--max-model-len',
+            '2048',
+            "--trust-remote-code",
+        ],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+    # Check the server output to confirm it's running
+    for line in process.stderr:
+      line = line.strip()
+      logger.info(line)
+      if "Uvicorn running on http://localhost:" in line:
+        self.server_started_event.set()
+        logger.info(f"vLLM Server started at http://localhost:{self.port}")
+        break
+
+  def predict(self, request: service_pb2.PostModelOutputsRequest
+             ) -> Iterator[service_pb2.MultiOutputResponse]:
+    """This is the method that will be called when the runner is run. It takes in an input and
+    returns an output.
+    """
+
+    # TODO: Could cache the model and this conversion if the hash is the same.
+    model = request.model
+    output_info = None
+    if request.model.model_version.id != "":
+      output_info = json_format.MessageToDict(
+          model.model_version.output_info, preserving_proto_field_name=True)
+
+    outputs = []
+    # TODO: parallelize this over inputs in a single request.
+    for inp in request.inputs:
+      output = resources_pb2.Output()
+
+      data = inp.data
+
+      # Optional use of output_info
+      inference_params = {}
+      if "params" in output_info:
+        inference_params = output_info["params"]
+
+      messages = []
+      temperature = inference_params.get("temperature", 0.7)
+      max_tokens = inference_params.get("max_tokens", 100)
+      top_p = inference_params.get("top_p", 1.0)
+
+      prompt = None
+      image_url = None
+      image_bytes = None
+      if data.text.raw != "":
+        prompt = data.text.raw
+      if data.image.url != "":
```

Comment:
i thought with the ensure urls downloaded we would always have base64 at this point? 

good to try calling this as a local dev runner as well through the API, it should do the same giving bytes to the model.

Reply:
Yeah we alway `base64`, we can remove `data.image.url ` part